VERSIONING.md

```diff
@@ -0,0 +1,104 @@
+[Obsolete](https://docs.microsoft.com/dotnet/api/system.obsoleteattribute)
+first.
+
+## Pre-releases
+
+Pre-release packages are denoted by appending identifiers such as
+-Alpha, -Beta, -RC
+etc. There are no API guarantees in pre-releases. In general, an RC pre-release
+is more stable than a Beta release, which is more stable than an Alpha release.
+
+### Public API change detection
+
+For convenience, every project in this repo uses
+[PublicApiAnalyzers](https://github.com/dotnet/roslyn-analyzers/tree/master/src/PublicApiAnalyzers)
+and lists public API in the directory "publicAPI". Any changes to public API,
+without corresponding changes here will result in build breaks - this helps
+catch any unintended changes to public API from being shipped accidentally. This
+also helps reviewers quickly understand if a given PR is making public API
+changes. For example,
+[this](https://github.com/open-telemetry/opentelemetry-dotnet/tree/master/src/OpenTelemetry.Instrumentation.AspNetCore/.publicApi)
+shows the public per target framework for the
+`OpenTelemetry.Instrumentation.AspNetCore` package.
+
+Since no stable version has been released so far, every API is listed in the
+"Unshipped.txt" file. Once 1.0.0 is shipped, it'll be moved to "Shipped.txt"
+file.
+
+## Packaging
+
+OpenTelemetry is structured around signals like Traces, Metrics, Logs, Baggage
+etc. OpenTelemetry .NET does not ship separate packages per signal. There is a
+single package which includes all the signals. i.e `OpenTelemetry.API` package
+will consist of API components from *all* the signals. `OpenTelemetry` package
+will consist of SDK components from *all* the signals. Instrumentations also
+follow the same model - for example, There will be a single package
+`OpenTelemetry.Instrumentation.AspNetCore` for ASP.NET Core instrumentation,
+which produces Traces, Metrics and handles propagation, instead of separate
+packages for traces and metrics.
+
+## Experimental Signals
+
+### Current approach
+
+Given the fact that all signals are shipped as a single package, the following
+approach is currently used to deal with experimental signals.
+
+Any experimental signal which is shipped as part of a normal package will be
+marked
+[Obsolete](https://docs.microsoft.com/dotnet/api/system.obsoleteattribute).
+(Ideally we need an "Experimental" attribute, but .NET has only built-in support
+for "Obsolete", so we are leveraging it. Using "Obsolete" code results in
+compile time warnings to caution users.) This allows an experimental signal to
+co-exist in the same package as other, non-experimental signals. For example,
+the OpenTelemetry 1.0.0 package will consist of Traces and Metrics. As Metrics
+signal is not ready for stable release, every method dealing with Metrics will
+be marked "Obsolete". Once Metrics signal gets stable, it'll be added to a
+release with minor version bump, say 1.2.0. The "Obsolete" Metric methods can be
```

Comment:
One tricky thing is that you may need two policies for `Deprecated` elements;

1) Ones using it as `Experimental` will have this attribute removed in the (near) future.
2) Ones that are effectively `Deprecated` and will be removed at some point in the (distant) future.

Reply:
@cijothomas Thanks for taking the time to write this up! I think I'm leaning towards approach 2 for this reason @carlosalberto brings up above. I'm thinking it will be hard for users to know from `[Obsolete]` what is being removed and what is experimental. We could help with the text in the attribute, but I think having the experimental stuff in a pre-release package is more obvious and a better experience for users than combing through warnings. Also, NuGet does support deprecated packages (example [Microsoft.CodeAnalysis.FxCopAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers) - notice the yellow warning, click for details) which we could possibly use when experimental stuff becomes official and is moved into main libraries?

Reply:
I think using `[Obsolete]` attribute for experimental features could be confusing, and is probably against the common practice?

I would vote for using `[Obsolete]` only on things that we plan to remove in the next major release, *not* on things that we plan to introduce in the future.